test/test_euler.py

```diff
@@ -0,0 +1,781 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import numpy.linalg as la  # noqa
+import pyopencl as cl
+import pyopencl.clrandom
+import pyopencl.clmath
+import logging
+import pytest
+
+from pytools.obj_array import (
+    flat_obj_array,
+    make_obj_array,
+)
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.symbolic.primitives import TracePair
+from mirgecom.euler import inviscid_operator
+from mirgecom.euler import split_conserved
+from mirgecom.steppers import euler_flow_stepper
+from mirgecom.initializers import Vortex2D
+from mirgecom.initializers import Lump
+from mirgecom.boundary import PrescribedBoundary
+from mirgecom.boundary import DummyBoundary
+from mirgecom.eos import IdealSingleGas
+from grudge.eager import EagerDGDiscretization
+from pyopencl.tools import (  # noqa
+    pytest_generate_tests_for_pyopencl as pytest_generate_tests,
+)
+
+
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_inviscid_flux(ctx_factory, dim):
+    """Checks that the Euler-internal inviscid flux
+    routine (_inviscid_flux) returns exactly the
+    expected result. The test checks flux for 1,
+    2, and 3 spatial dimensions.
+
+    The test is in 3 parts:
+
+    - Exact expression: Checks the returned flux
+      against the exact expressions:
+      F(q) = <rhoV, (E+p)V, rho(V.x.V) + pI>
+    - Prescribed p, V = 0:  Checks that only
+      diagonal terms of the momentum flux
+      [ rho(V.x.V) + pI ] are non-zero and return
+      the correctly calculated p.
+    - Prescribed p, V != 0: Checks that the
+      flux terms are returned in the proper
+      order by running only 1 non-zero velocity
+      component at-a-time.
+    """
+    cl_ctx = ctx_factory()
+    queue = cl.CommandQueue(cl_ctx)
+    logger = logging.getLogger(__name__)
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    #    for dim in [1, 2, 3]:
+    mesh = generate_regular_rect_mesh(
+        a=(-0.5,) * dim, b=(0.5,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+    discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+    eos = IdealSingleGas()
+
+    logger.info(f"Number of {dim}d elems: {mesh.nelements}")
+
+    mass = cl.clrandom.rand(
+        queue, (mesh.nelements,), dtype=np.float64
+    )
+    energy = cl.clrandom.rand(
+        queue, (mesh.nelements,), dtype=np.float64
+    )
+    mom = make_obj_array(
+        [
+            cl.clrandom.rand(
+                queue, (mesh.nelements,), dtype=np.float64
+            )
+            for i in range(dim)
+        ]
+    )
+
+    q = flat_obj_array(mass, energy, mom)
```

Comment:
Reading this (and thinking of all the other places with code like this) makes me think that we should probably also have the "inverse" (in a sense) of `split_conserved`, so that we can avoid hardcoding the state vector ordering more or less entirely.

Reply:
I'm curious what this would look like.  Can we chat about it?

Reply:
Pretty simple. Maybe we want something like:
```python
class EulerSetup:
    def __init__(self, dim, eos):
        # It's tempting to use this as a grab bag of all sorts of stuff. IMO we should try to keep it minimal.
    def split_state_vec(self, q):
        # This replaces split_conserved
    def make_state_vec(self, mass, energy, mom):
        return flat_obj_array(...)
```
What do you think?